### PR TITLE
New 07 Registers

### PR DESCRIPTION
**Pull request by winksaville**
Modified to run using `stm32f3-discovery` instead of `f3` and updated
the `.md` files appropriately.

### CODE REVIEW

File: src/07-registers/bad-address.md [hunk 1]
```diff
@@ -29,16 +29,20 @@ Invalid in the sense that there's no register at this address.
...
+Breakpoint 1, registers::__cortex_m_rt_main_trampoline () at src/07-registers/src/main.rs:9
+9       #[entry]
 
 (gdb) continue
 Continuing.
 
-Breakpoint 2, cortex_m_rt::HardFault_ (ef=0x10001fb0)
-    at $REGISTRY/cortex-m-rt-0.6.13/src/lib.rs:560
+Breakpoint 3, cortex_m_rt::HardFault_ (ef=0x20009fb0)
+    at /home/wink/.cargo/registry/src/github.com-1ecc6299db9ec823/cortex-m-rt-0.6.13/src/lib.rs:560
```
Comment (eldruin) [inline src/07-registers/bad-address.md:42]:
```suggestion
    at ~/.cargo/registry/src/github.com-1ecc6299db9ec823/cortex-m-rt-0.6.13/src/lib.rs:560
```

File: src/07-registers/bad-address.md [hunk 2]
```diff
@@ -29,16 +29,20 @@ Invalid in the sense that there's no register at this address.
 Now, let's try it.
 
 ``` console
+ cargo run
```
Comment (eldruin) [inline src/07-registers/bad-address.md:32]:
```suggestion
```
The command appears twice.

File: src/07-registers/optimization.md [hunk 3]
```diff
@@ -183,32 +145,77 @@ fn main() -> ! {
...
+
 ```
 
-If we look at the disassembly of this new program compiled in release mode:
+Generate `release.txt` using with `--release` mode.
 
 ``` console
-$ cargo objdump --bin registers --release -- -d --no-show-raw-insn --print-imm-hex --source
-registers:      file format ELF32-arm-little
+cargo objdump --release --bin registers -- -d --no-show-raw-insn --print-imm-hex --source > release.txt
```
Comment (eldruin) [inline src/07-registers/optimization.md:154]:
```suggestion
$ cargo objdump --release --bin registers -- -d --no-show-raw-insn --print-imm-hex --source > release.txt
```
For consistency

File: src/07-registers/type-safe-manipulation.md [hunk 4]
```diff
@@ -43,13 +43,13 @@ fn main() -> ! {
     // Turn on the North LED
     gpioe.bsrr.write(|w| w.bs9().set_bit());
 
-    // Turn on the East LED
+    // Turn on the Eask LED
```
Comment (eldruin) [inline src/07-registers/type-safe-manipulation.md:46]:
```suggestion
    // Turn on the East LED
```

File: src/07-registers/type-safe-manipulation.md [hunk 5]
```diff
@@ -43,13 +43,13 @@ fn main() -> ! {
...
     gpioe.bsrr.write(|w| w.bs11().set_bit());
 
     // Turn off the North LED
     gpioe.bsrr.write(|w| w.br9().set_bit());
 
-    // Turn off the East LED
+    // Turn off the Eask LED
```
Comment (eldruin) [inline src/07-registers/type-safe-manipulation.md:52]:
```suggestion
    // Turn off the East LED
```

File: src/07-registers/type-safe-manipulation.md [hunk 6]
```diff
@@ -175,28 +192,38 @@ did!
...
- 8000194:       movt    r0, #0x4800
- 8000198:       str     r1, [r0]
- 800019a:       mov.w   r1, #0x800
- 800019e:       str     r1, [r0]
- 80001a0:       mov.w   r1, #0x2000000
- 80001a4:       str     r1, [r0]
- 80001a6:       mov.w   r1, #0x8000000
- 80001aa:       str     r1, [r0]
- 80001ac:       b       #-0x4 <main+0x24>
+cargo objdump --bin registers --release -- -d --no-show-raw-insn --print-imm-hex > release.txt
```
Comment (eldruin) [inline src/07-registers/type-safe-manipulation.md:197]:
```suggestion
$ cargo objdump --bin registers --release -- -d --no-show-raw-insn --print-imm-hex > release.txt
```

**Review by eldruin — COMMENTED** [review]
Other than a couple nitpics looks good to me. Thanks!

File: src/07-registers/bad-address.md (hunk 2 above)
Comment (winksaville) [reply src/07-registers/bad-address.md]: Done

File: src/07-registers/bad-address.md (hunk 1 above)
Comment (winksaville) [reply src/07-registers/bad-address.md]: Done

File: src/07-registers/optimization.md (hunk 3 above)
Comment (winksaville) [reply src/07-registers/optimization.md:154]: I purposely didn't add the `$` as this makes the `copy` button more useful.

File: src/07-registers/type-safe-manipulation.md (hunk 5 above)
Comment (winksaville) [reply src/07-registers/type-safe-manipulation.md]: Done

File: src/07-registers/type-safe-manipulation.md (hunk 4 above)
Comment (winksaville) [reply src/07-registers/type-safe-manipulation.md]: Done

File: src/07-registers/type-safe-manipulation.md (hunk 6 above)
Comment (winksaville) [reply src/07-registers/type-safe-manipulation.md:197]: I purposely didn't add the `$` as this makes the `copy` button more useful.

**Review by winksaville — COMMENTED** [review]
I've resolved these with the exception that I had purposely left off the `$`. If you prefer I can add them, but wanted to see if you agreed or not. If think making the copy useful I'll continue with that technique moving forward and change similar commands in 05 and 06 if they need updating. Note; I'll be updating 05 and 06 anyway as right now they are using stm32f3-discovery v0.5.0 and I'm using v0.6.0 here. This was needed to get the RegisterBlock stuff working in "7.5 Type safe manipulation".

**Review by eldruin — DISMISSED** [review]
I see. That about the `$` is a good point. Let's get rid of it where it makes sense, then.
Thank you for your work!

bors r+

**Comment by winksaville** [comment]
> I see. That about the `$` is a good point. Let's get rid of it where it makes sense, then.
> Thank you for your work!
> 

I'll update 05 and 06 to use `stm32f3-discovery v0.6.0` tomorrow and remove `$` where appropriate.

File: src/07-registers/auxiliary/src/lib.rs [hunk 7]
```diff
@@ -3,28 +3,38 @@
 #![deny(warnings)]
 #![no_std]
 
-#[allow(unused_extern_crates)] // NOTE(allow) bug rust-lang/rust#53964
 extern crate panic_itm; // panic handler
```
Comment (eldruin) [inline src/07-registers/auxiliary/src/lib.rs:6]:
```suggestion
use panic_itm as _; // panic handler
```
This fails on CI otherwise due to old idioms

Comment (winksaville) [reply src/07-registers/auxiliary/src/lib.rs:6]: You're suggesting `use panic_item as _;` but that idiom isn't used anywhere else,
instead `pub use panic_itm;` is used:

```
$ find . -type f -name '*.rs' | xargs grep -i 'panic_itm'
./book/08-leds-again/auxiliary/src/lib.rs:extern crate panic_itm; // panic handler
./book/07-registers/auxiliary/src/lib.rs://pub use panic_itm; // panic handler
./book/07-registers/auxiliary/src/lib.rs:use panic_itm as _; // panic handler
./book/05-led-roulette/auxiliary/src/lib.rs:pub use panic_itm; // panic handler
./book/06-hello-world/auxiliary/src/lib.rs:pub use panic_itm;
./book/14-i2c/auxiliary/src/lib.rs:extern crate panic_itm; // panic handler
./book/09-clocks-and-timers/auxiliary/src/lib.rs:extern crate panic_itm; // panic handler
./book/11-usart/auxiliary/src/lib.rs:extern crate panic_itm; // panic handler
./book/15-led-compass/auxiliary/src/lib.rs:extern crate panic_itm; // panic handler
./book/16-punch-o-meter/auxiliary/src/lib.rs:extern crate panic_itm; // panic handler
./src/08-leds-again/auxiliary/src/lib.rs:extern crate panic_itm; // panic handler
./src/07-registers/auxiliary/src/lib.rs:use panic_itm as _; // panic handler
./src/05-led-roulette/auxiliary/src/lib.rs:pub use panic_itm; // panic handler
./src/06-hello-world/auxiliary/src/lib.rs:pub use panic_itm;
./src/14-i2c/auxiliary/src/lib.rs:extern crate panic_itm; // panic handler
./src/09-clocks-and-timers/auxiliary/src/lib.rs:extern crate panic_itm; // panic handler
./src/11-usart/auxiliary/src/lib.rs:extern crate panic_itm; // panic handler
./src/15-led-compass/auxiliary/src/lib.rs:extern crate panic_itm; // panic handler
./src/16-punch-o-meter/auxiliary/src/lib.rs:extern crate panic_itm; // panic handler
```
I assume I should change all:
```
#[allow(unused_extern_crates)] // NOTE(allow) bug rust-lang/rust#53964
extern crate panic_itm; // panic_handler
```
To either `pub use panic_itm;` or `use panic_itm as _;`?

Comment (winksaville) [reply src/07-registers/auxiliary/src/lib.rs:6]: After reading [use-declarations](https://doc.rust-lang.org/reference/items/use-declarations.html) I'm going to go ahead with `use panic_itm as _;`.  It seems the best choice as there seems to be no reason to re-export by making it public, but we have to indicate that its the `panic_handler` we want. As I make modications moving forward I'll change all occurances to be `use panic_itm as _;`.

**Review by eldruin — APPROVED** [review]
Thanks @winksaville!

bors r+

**Pull request merged**